src/fl/federation/strategy.py

```diff
@@ -33,56 +36,46 @@ def on_evaluate_config_fn(rnd: int):
 
 
 class MlflowLogger:
-    def __init__(self, epochs_in_round: int) -> None:
+    def __init__(self, epochs_in_round: int, model_type: str) -> None:
         """Logs server-side per-round metrics to mlflow
         """        
         self.epochs_in_round = epochs_in_round
+        if model_type not in ['lassonet_regressor', 'mlp_regressor']:
+            raise ValueError(f'model_type should be one of the lassonet_regressor, mlp_regressor and not {model_type}')
 
-    def _calculate_agg_metric(self, metric_name: str, results: RESULTS, custom_len_name: str = None) -> float:
-        """Calculates weighted average {metric_name} from {results}
+        self.model_type = model_type
 
-        Args:
-            metric_name (str): Name of metric in {results}
-            results (RESULTS): List of client's client proxies and current round metrics
-            custom_len_name (str): Needed if we are calculating test metrics using test_len value returned from each client
-        Returns:
-            float: averaged metric
-        """        
-        if custom_len_name is None:
-            losses = [r.metrics[metric_name] * r.num_examples for _, r in results]
-            examples = [r.num_examples for _, r in results]
+    def _get_reduction_type(self, rnd: int) -> str:
+        if self.model_type == 'lassonet_regressor':
+            return 'mean' if rnd == -1 else 'lassonet_best'
         else:
-            losses = [r.metrics[metric_name] * r.metrics[custom_len_name] for _, r in results]
-            examples = [r.metrics[custom_len_name] for _, r in results]
-
-        return sum(losses) / sum(examples)
+            return 'mean' 
 
-    def log_losses(self, rnd: int, results: RESULTS) -> float:
+    def log_losses(self, rnd: int, results: RESULTS) -> Metrics:
         """Logs val_loss, val and train r^2 to mlflow
 
         Args:
             rnd (int): current FL round
             results (RESULTS): Central model evaluation results from server
 
         Returns:
-            float: Current validation loss
-        """        
-        train_loss = self._calculate_agg_metric('train_loss', results)
-        val_loss = self._calculate_agg_metric('val_loss', results)
-        train_r2 = self._calculate_agg_metric('train_r2', results)
-        val_r2 = self._calculate_agg_metric('val_r2', results)
-        logging.info(f"round {rnd}\ttrain_loss: {train_loss:.4f}\ttrain_r2: {train_r2:.4f}\tval_loss: {val_loss:.2f}\tval_r2: {val_r2:.4f}")
-        mlflow.log_metric('train_loss', train_loss, step=rnd*self.epochs_in_round)
-        mlflow.log_metric('val_loss', val_loss, step=rnd*self.epochs_in_round)
-        mlflow.log_metric('train_r2', train_r2, step=rnd*self.epochs_in_round)
-        mlflow.log_metric('val_r2', val_r2, step=rnd*self.epochs_in_round)
-
-        if rnd == -1:
+            Metrics: Metrics reduced over clients
+        """       
+        metric_list = [pickle.loads(r[1].metrics['metrics']) for r in results]
```

Comment:
Flower can pass only str, int, float, bytes objects between nodes and a server. We cannot store them as text files in a federated setting. LassoNetMetrics is a relatively complex object and the only alternative to pickle will be json.loads\dumps. In case of json, we will be forced to choose the correct metric class ourselves.